### PR TITLE
opentracing-shim: support span wrappers

### PR DESCRIPTION
**Pull request by zeitlinger**
make the tracing shim compatible with span wrappers, such as https://github.com/zalando/opentracing-toolbox/tree/main/opentracing-proxy

### CODE REVIEW

**Comment by jkwatson** [comment]
can you add unit tests for this change? Thanks!

File: opentracing-shim/src/main/java/io/opentelemetry/opentracingshim/SpanBuilderShim.java [hunk 1]
```diff
@@ -237,17 +238,26 @@ public Span start() {
...
   }
 
-  private static SpanShim getSpanShim(Span span) {
+  static SpanShim getSpanShim(Span span) {
     if (!(span instanceof SpanShim)) {
-      throw new IllegalArgumentException("span is not a valid SpanShim object");
+      if (span instanceof Supplier<?>) {
+        // allow libraries to implement a delegate span,
+        // such as https://github.com/zalando/opentracing-toolbox/tree/main/opentracing-proxy
+        return getSpanShim((Span) ((Supplier<?>) span).get());
```
[the remark this answers is not in the document]
Comment (zeitlinger) [reply opentracing-shim/src/main/java/io/opentelemetry/opentracingshim/SpanBuilderShim.java]: good idea - done

File: opentracing-shim/src/test/java/io/opentelemetry/opentracingshim/SpanShimTest.java [hunk 2]
```diff
@@ -190,6 +191,25 @@ void log_fields_micros() {
...
+  }
+
+  @Test
+  void getContextShim() {
+    SpanContextShim contextShim = new SpanContextShim(new SpanShim(telemetryInfo, span));
+    assertThat(SpanBuilderShim.getContextShim(contextShim)).isEqualTo(contextShim);
+    assertThatThrownBy(() -> SpanBuilderShim.getContextShim(null))
+        .hasMessage("context is not a valid SpanContextShim object: null");
+  }
+
```
Comment (jkwatson) [inline opentracing-shim/src/test/java/io/opentelemetry/opentracingshim/SpanShimTest.java:212]: let's add a test to verify that the negative case works, as well (i.e. some other kind of Supplier).

**Comment by jkwatson** [comment]
@carlosalberto any reason this isn't a good idea?

File: opentracing-shim/src/test/java/io/opentelemetry/opentracingshim/SpanShimTest.java (hunk 2 above)
Comment (zeitlinger) [reply opentracing-shim/src/test/java/io/opentelemetry/opentracingshim/SpanShimTest.java:212]: the context shim doesn't have a supplier, only the span shim - test case for that is here: https://github.com/open-telemetry/opentelemetry-java/pull/4535/files#diff-f628731aad63f58834336eb865007007c95db889f79cb71adc2e0a0072b297c5R199

**Review by jkwatson — APPROVED** [review]

**Review by carlosalberto — APPROVED** [review]
